### PR TITLE
[CNXC-193] Multi-user functionality

### PR DESCRIPTION
**Pull request by janakitti**
### Story Description
COVID-Net UI should support multiple user accounts (leveraging the ChRIS backend). Currently, the UI has a single user hardcoded into the implementation (chris1234); this should be updated to dynamically use the current user logged into the UI.

### Implementation
- Remove all hard-coded user data.
- Remove password property from user store
- Ensure that `ChrisAPIClient tokenIsUnauthorized` is set to `true` when user logs out to prevent auth token from being carried over across multiple sessions.
- Move up code for checking if user is already authenticated to new RouterWrapper component
- Automatically redirect from Login to Dashboard if user already authenticated

### CODE REVIEW

**Review by boris-nguyen — COMMENTED** [review]
Hey @janakitti 

Overall, good work. I added some minor comments, which I'm sure you'd be able to fix quickly.

One thing that would be great to resolve (if time permits) is where should we put the logic to authenticate users that have already been authenticated in the session. The placement of that code inside PageWrapper indeed works. However, since it wraps all pages, whenever a user navigates to a different page in COVIDNet, it triggers the auth related code to run so unnecessary HTTP requests are being made (such as client.getUser()).

File: src/api/chrisapiclient.ts [hunk 1]
```diff
@@ -17,15 +17,15 @@ const AUTH_TOKEN_KEY = 'AUTH_TOKEN';
 class ChrisAPIClient {
 
   private static client: Client;
-  private static tokenUnauthorized: boolean;
+  private static tokenIsUnauthorized: boolean;
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/api/chrisapiclient.ts]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/ed2ff013f954ec45119fe0285fc3343a3837b901

File: src/api/chrisapiclient.ts [hunk 2]
```diff
@@ -34,6 +34,10 @@ class ChrisAPIClient {
     return this.client;
   }
 
+  static setTokenIsUnauthorized(tokenIsUnauthorized: boolean) {
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/api/chrisapiclient.ts]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/ed2ff013f954ec45119fe0285fc3343a3837b901

File: src/containers/Layout/PageWrapper.tsx [hunk 3]
```diff
@@ -13,19 +15,19 @@ interface WrapperProps {
...
+    const token = window.sessionStorage.getItem('AUTH_TOKEN');
     if (!!token) {
-      dispatch({
-        type: Types.Login_update,
-        payload: {
-          username: 'chris',
-          password: 'chris1234'
-        }
-      });
+      const client: any = ChrisAPIClient.getClient();
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/containers/Layout/PageWrapper.tsx]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/abe8a0f95f23ab31053dec014dd75915428ad709

File: src/containers/Layout/PageWrapper.tsx [hunk 4]
```diff
@@ -13,19 +15,19 @@ interface WrapperProps {
...
+      client.getUser().then((res: UserResponse) => {
+        const user: IUserState = res?.data;
+        dispatch({
+          type: Types.Login_update,
+          payload: user,
+        });
+      })
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/containers/Layout/PageWrapper.tsx]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/0c61cb9aab0fd34b2d696a02cefa8a83ad865cf6

File: src/services/login.tsx [hunk 5]
```diff
@@ -2,17 +2,12 @@ import Client from "@fnndsc/chrisapi";
 import { IUserState } from '../context/reducers/userReducer';
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/services/login.tsx]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/c8ce89ccd152a1e2041335bcaf0ee96a12674e80

File: src/services/login.tsx [hunk 6]
```diff
@@ -2,17 +2,12 @@ import Client from "@fnndsc/chrisapi";
...
 
 // returns true if login succeed and false otherwises
```
[the remark this answers is not in the document]
Comment (janakitti) [reply src/services/login.tsx]: Done: https://github.com/darwinai/covidnet_ui/pull/22/commits/c8ce89ccd152a1e2041335bcaf0ee96a12674e80

File: src/api/chrisapiclient.ts [hunk 7]
```diff
@@ -17,15 +17,15 @@ const AUTH_TOKEN_KEY = 'AUTH_TOKEN';
 class ChrisAPIClient {
 
   private static client: Client;
-  private static tokenUnauthorized: boolean;
+  private static isTokenAuthorized: boolean;
 
   static getClient(): Client {
-    if (!this.client || this.tokenUnauthorized) {
+    if (!this.client || !this.isTokenAuthorized) {
```
Comment (colinywu) [inline src/api/chrisapiclient.ts:23]: how come there is a negation `!` here, when it was not used previously?

File: src/containers/Layout/Header.tsx [hunk 8]
```diff
@@ -26,14 +27,15 @@ const Header: React.FC<HeaderProps> = ({ onNotificationBadgeClick }) => {
   const history = useHistory()
   const { state, dispatch } = React.useContext(AppContext);
   const [isDropdownOpen, setIsDropdownOpen] = React.useState(false)
+  const username = state.user.username;
 
   const onDropdownSelect = () => {
     setIsDropdownOpen(!isDropdownOpen)
```
Comment (colinywu) [inline src/containers/Layout/Header.tsx:33]: semi-colon

File: src/api/chrisapiclient.ts (hunk 7 above)
Comment (janakitti) [reply src/api/chrisapiclient.ts:23]: The variable `tokenUnauthorized` was renamed to `isTokenAuthorized`, and so we needed to negate it.

File: src/RouterWrapper.tsx [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+          type: Types.Login_update,
+          payload: user
+        });
+        
+        // If on Login page and already authenticated, go to Dashboard
+        if (location.pathname === "/login") {
+          history.push('/');
+        }
+      }
+      setRenderChildren(true);
```
Comment (colinywu) [inline src/RouterWrapper.tsx:36]: Would we want this inside the `if` statement, since we only want the value to be true, if there's a valid token?

**Review by colinywu — APPROVED** [review]
Great work Janakitti! Tested it out, and everything works great (including the `RouterWrapper` fix).

File: src/RouterWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        const res: User = await client.getUser();
+        const user: IUserState = res?.data;
+        dispatch({
+          type: Types.Login_update,
+          payload: user
+        });
+        
+        // If on Login page and already authenticated, go to Dashboard
+        if (location.pathname === "/login") {
+          history.push('/');
```
Comment (colinywu) [inline src/RouterWrapper.tsx:33]: How come you need to route them from the RouterWRapper? How was previous the user routed to the Dashboard previously?

**Review by boris-nguyen — APPROVED** [review]

File: src/RouterWrapper.tsx (hunk 9 above)
Comment (boris-nguyen) [reply src/RouterWrapper.tsx:36]: If there is no token, we proceed with rendering the children. The reason behind not rendering right away is we needed to ensure tokens were checked before the RouteGuard checks whether the user has already logged in or not.

File: src/RouterWrapper.tsx (hunk 10 above)
Comment (boris-nguyen) [reply src/RouterWrapper.tsx:33]: Previously, if auth token already existed, a user would still need to manually log into COVIDNet. The routing above redirects a user to the Dashboard if a token already exists.

**Review by IvanToronto — APPROVED** [review]

**Pull request merged**